File: app/src/main/java/me/yxp/qfun/utils/qq/HostInfo.java
package me.yxp.qfun.utils.qq;

import android.content.Context;
import android.content.pm.PackageInfo;
import android.content.pm.PackageManager;

public class HostInfo {
    public static final String PACKAGE_NAME_QQ = "com.tencent.mobileqq";
    public static final String PACKAGE_NAME_TIM = "com.tencent.tim";

    private static long sVersionCode;
    private static String sModuleDataPath;
    private static Context sHostContext;

    public static void setHostInfo(Context context) {
        setHostContext(context);
        setVersionCode(context);
        setModuleDataPath(context);
    }

    public static long getVersionCode() {
        return sVersionCode;
    }

    private static void setVersionCode(Context context) {
        try {
            PackageInfo packageInfo = context.getPackageManager().getPackageInfo(context.getPackageName(), 0);
            sVersionCode = packageInfo.getLongVersionCode();
        } catch (PackageManager.NameNotFoundException ignored) {
        }
    }

    public static String getPackageName() {
        return sHostContext.getPackageName();
    }

    public static boolean isQQ() {
        return PACKAGE_NAME_QQ.equals(getPackageName());
    }

    public static boolean isTIM() {
        return PACKAGE_NAME_TIM.equals(getPackageName());
    }

    public static String getMODULE_DATA_PATH() {
        return sModuleDataPath;
    }

    private static void setModuleDataPath(Context context) {
        sModuleDataPath = context.getExternalFilesDir(null).getParent() + "/QFun/";
    }

    public static Context getHostContext() {
        return sHostContext;
    }

    private static void setHostContext(Context context) {
        sHostContext = context;
    }

    public static boolean isInHostProcess() {
        if (sHostContext == null) return false;
        String packageName = getPackageName();
        return PACKAGE_NAME_QQ.equals(packageName) || PACKAGE_NAME_TIM.equals(packageName);
    }
}

File: app/src/main/java/me/yxp/qfun/utils/qq/HostInfo.kt
package me.yxp.qfun.utils.qq

import android.annotation.SuppressLint
import android.content.Context
import android.os.Build

@Suppress("DEPRECATION")
@SuppressLint("StaticFieldLeak")
object HostInfo {
    const val PACKAGE_NAME_QQ = "com.tencent.mobileqq"
    const val PACKAGE_NAME_TIM = "com.tencent.tim"

    lateinit var hostContext: Context
        private set

    var packageName = ""

    var hostName: String = "Host App"
        private set

    lateinit var processName: String

    val isQQ: Boolean
        get() = packageName == PACKAGE_NAME_QQ
    val isTIM: Boolean
        get() = packageName == PACKAGE_NAME_TIM

    val isInHostProcess: Boolean
        get() = isQQ || isTIM
    var versionCode: Long = 0
        private set
    var versionName: String = ""
        private set
    var moduleDataPath: String = ""
        private set

    @JvmStatic
    fun init(context: Context) {
        hostContext = context

        runCatching {
            val pm = context.packageManager
            val info = pm.getPackageInfo(context.packageName, 0)
            versionCode =
                if (Build.VERSION.SDK_INT > 28)
                    info.longVersionCode
                else
                    info.versionCode.toLong()
            versionName = info.versionName.orEmpty()
            val appInfo = pm.getApplicationInfo(packageName, 0)
            hostName = pm.getApplicationLabel(appInfo).toString()
        }

        val externalDir = context.getExternalFilesDir(null)?.parentFile
        moduleDataPath = externalDir?.let { "${it.absolutePath}/QFun/" }
            ?: "/storage/emulated/0/Android/data/${context.packageName}/QFun/"
    }

}